src/anyio/_core/_synchronization.py

```diff
@@ -302,6 +304,8 @@ async def acquire(self) -> None:
             except BaseException:
                 if not event.is_set():
                     self._waiters.remove(event)
+                else:
+                    self._value += 1
```

Comment:
I'm thinking, maybe this should call `self.release()` too, like the `Lock` version?